actors/builtin/miner/miner_actor.go

```diff
@@ -481,6 +461,185 @@ func (a Actor) SubmitWindowedPoSt(rt Runtime, params *SubmitWindowedPoStParams)
 	return nil
 }
 
+type ChallengeWindowedPoStParams struct {
+	Deadline   uint64
+	ProofIndex uint64 // only one is allowed at a time to avoid loading too many sector infos.
+}
+
+func (a Actor) ChallengeWindowedPoSt(rt Runtime, params *ChallengeWindowedPoStParams) *abi.EmptyValue {
+	rt.ValidateImmediateCallerAcceptAny()
+
+	if params.Deadline >= WPoStPeriodDeadlines {
+		rt.Abortf(exitcode.ErrIllegalArgument, "invalid deadline %d of %d", params.Deadline, WPoStPeriodDeadlines)
+	}
+
+	currEpoch := rt.CurrEpoch()
+
+	// TODO: maybe stash this in the deadline for better accuracy?
+	epochReward := requestCurrentEpochBlockReward(rt)
+	pwrTotal := requestCurrentTotalPower(rt)
+
+	toBurn := abi.NewTokenAmount(0)
+	pledgeDelta := abi.NewTokenAmount(0)
+	powerDelta := NewPowerPairZero()
+	var st State
+	rt.StateTransaction(&st, func() {
+		if st.CurrentDeadline == params.Deadline {
+			// TODO: blank out the previous deadline as well?
+			rt.Abortf(exitcode.ErrForbidden, "cannot challenge a window post for an open deadline")
+		}
+
+		info := getMinerInfo(rt, &st)
+		faultyPower := NewPowerPairZero()
+		store := adt.AsStore(rt)
+
+		// Check proof
+		// TODO: move into deadline state function.
+		{
+			// Load the target state.
+			deadlinesCurrent, err := st.LoadDeadlines(store)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadlines")
+
+			dlCurrent, err := deadlinesCurrent.LoadDeadline(store, params.Deadline)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load deadline")
+
+			proofsSnapshot, err := adt.AsArray(store, dlCurrent.PoStSubmissionsSnapshot, DeadlinePoStSubmissionsAmtBitwidth)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load proofs")
+
+			partitionsSnapshot, err := adt.AsArray(store, dlCurrent.PartitionsSnapshot, DeadlinePoStSubmissionsAmtBitwidth)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load partitions")
+
+			// Load the target proof.
+			var post WindowedPoSt
+			found, err := proofsSnapshot.Get(params.ProofIndex, &post)
+			builtin.RequireNoErr(rt, err, exitcode.ErrIllegalState, "failed to load proof %d", params.ProofIndex)
+			if !found {
+				rt.Abortf(exitcode.ErrIllegalArgument, "failed to find post %d", params.ProofIndex)
+			}
+
+			var allSectors, allIgnored []bitfield.BitField
+			faults := make(PartitionSectorMap)
+			err = post.Partitions.ForEach(func(partIdx uint64) error {
+				var partitionSnapshot Partition
+				if found, err := partitionsSnapshot.Get(partIdx, &partitionSnapshot); err != nil {
+					return err
+				} else if !found {
+					return exitcode.ErrIllegalState.Wrapf("failed to find partition when challenging proof %d", params.ProofIndex)
+				}
+
+				// Record sectors for proof verification
+				allSectors = append(allSectors, partitionSnapshot.Sectors)
+				allIgnored = append(allIgnored, partitionSnapshot.Faults)
+				allIgnored = append(allIgnored, partitionSnapshot.Terminated)
```

Comment:
Because of the logic in `ConfirmSectorProofsValid`, `partitionSnapshot` should absolutely NOT have any `Unproven` sectors.

On the off chance it does, shouldn't they be ignored?

Reply:
You're right, they should. It won't hurt much to just add them in.